### PR TITLE
Simple service-worker

### PR DESCRIPTION
**Pull request by Nhowka**
Caches all requests and respond with the cached response, except the ones starting with '/samples' where it's network then cache so the samples are always fresh.

### CODE REVIEW

**Comment by MangelMaxime** [comment]
Will this detect changes in the dlls ? And relief them ?

**Comment by Nhowka** [comment]
At the moment no, I'll check if there is a way to use the network only when newer.

**Comment by MangelMaxime** [comment]
If we need to add an extra file or variable somewhere to store the version we can do it

**Comment by Nhowka** [comment]
There is an option to load from cache and update on the background. It will load the old file once but load the new one on the next run. I think it's the easiest to implement and could work well.

**Comment by MangelMaxime** [comment]
Hum, I would be ok if we can inform the user that he "need" to reload it's tab to get the latest versions.

Otherwise, I can see some issue like: 

- "You said it was fixed but it's not for me"
- You need to reload
- "Ah ok"
-etc. :)

File: src/App/Main.fs [hunk 1]
```diff
@@ -524,6 +524,8 @@ let private subscriptions (model: Model) =
 
     Cmd.ofSub sub
 
+Browser.navigator.serviceWorker.register("/service-worker.js") |> ignore
```
Comment (alfonsogarciacaro) [inline src/App/Main.fs]: Do we need to do this in the worker code (it's the worker which does most of the downloads) as well, or is it enough to just put it here?

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for this @Nhowka! Dlls shouldn't change very often, but the three `Fable.*` ones may do from time to time. I guess we can just bump the version of `fableCache` to invalidate the cache. Although I'd like to unify the way of invalidating caches to make maintenance easier. At the moment we have:

- For the JS file, Webpack just adds a hash to the file so every time it changes we are sure the browser won't download a cached version (though this is not true for `bundle.min.js` which contains the biggest chunk of JS code).
- We also have the [Fable version number](https://github.com/fable-compiler/repl2/blob/c13bdc4083696230de82e6ad5617ae4394bf2330/src/App/Shared.fs#L12), though atm this is just informative and probably cannot be used to invalidate the cache either, because it may happen that Fable is not updated but the dlls are.
- Now, `fableCache` in the service-worker file.

It may be a good option to unify all of the above in a single cache version (that is placed together with the other literals in the Shared.fs file) and let the build script update it every time we change anything that is not the samples. (I would add a new FAKE target for publishing new samples without touching the cache version.) What do you think?

File: src/App/Main.fs (hunk 1 above)
Comment (Nhowka) [reply src/App/Main.fs]: Just here is enough, the service worker has a wide scope

**Comment by Nhowka** [comment]
Having them unified should be a good idea, invalidating the cache on a big change is safer and the download isn't thaaaat big.

**Comment by MangelMaxime** [comment]
For bundle.js we just need to use bundle.[hash].js in webpack config. 

I am on mobile so i can’t format correctly 😊

**Comment by alfonsogarciacaro** [comment]
@MangelMaxime It's a bit more complicated, because `bundle.min.js` is stored as an Appveyor artifact after a successful build in the Fable repo. I guess we could still attach the hash and in the build script of this repo, infer the hash and [change this line](https://github.com/fable-compiler/repl2/blob/c13bdc4083696230de82e6ad5617ae4394bf2330/src/App/Shared.fs#L14). But if we're to use the service worker, I guess it's better to use the cache version instead of managing too many hashes. It should be OK to just download everything anew when there's a change, the biggest time at start up is actually the creation of the F# compiler checker.

Anyways, let's merge this PR and I'll try to edit the build script when I have a moment. Thanks again for this @Nhowka!

**Comment by alfonsogarciacaro** [comment]
I just read service workers only work over https, but we don't have it activated yet for fable.io. Will this be a problem?

**Comment by Nhowka** [comment]
Oh, yeah... That will be a problem. I believe that the browser will just refuse to register the service worker, so it will just work as before.

**Pull request merged**